File: src/main/java/com/baulsupp/oksocial/completion/CompletionMappings.java
package com.baulsupp.oksocial.completion;

import com.google.common.collect.Lists;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.function.Function;
import java.util.function.Supplier;

import static java.util.concurrent.CompletableFuture.completedFuture;

public class CompletionMappings {
  private List<Function<UrlList, CompletableFuture<UrlList>>> mappings = Lists.newArrayList();

  public void withVariable(String name, List<String> values) {
    withVariable(name, () -> completedFuture(values));
  }

  public void withVariable(String name, Supplier<CompletableFuture<List<String>>> values) {
    mappings.add(ul -> values.get().thenApply(l -> ul.replace(name, l, true)));
  }

  public CompletableFuture<UrlList> replaceVariables(UrlList urlList) {
    CompletableFuture<UrlList> future = CompletableFuture.completedFuture(urlList);

    for (Function<UrlList, CompletableFuture<UrlList>> s : mappings) {
      future = future.thenCompose(s);
    }

    return future;
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/completion/CompletionMappings.kt
package com.baulsupp.oksocial.completion

import io.github.vjames19.futures.jdk8.ImmediateFuture
import io.github.vjames19.futures.jdk8.flatMap
import io.github.vjames19.futures.jdk8.map
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletableFuture.completedFuture

class CompletionMappings {
    private val mappings = mutableListOf<(UrlList) -> CompletableFuture<UrlList>>()

    fun withVariable(name: String, values: List<String>) {
        withVariable(name, { completedFuture(values) })
    }

    fun withVariable(name: String, values: () -> CompletableFuture<List<String>>) {
        val element = { ul: UrlList -> values().map { l -> ul.replace(name, l, true) } }
        mappings.add(element)
    }

    fun replaceVariables(urlList: UrlList): CompletableFuture<UrlList> {
        var future = ImmediateFuture { urlList }

        for (s in mappings) {
            future = future.flatMap { s(it) }
        }

        return future
    }
}
